internal/logging/panic.go

```diff
@@ -48,6 +48,39 @@ func PanicHandler() {
 	defer panicMutex.Unlock()
 
 	recovered := recover()
+	panicHandler(recovered, nil)
+}
+
+// PanicHandlerWithTraceFn returns a function similar to PanicHandler which is
+// called to recover from an internal panic in OpenTofu, and augments the
+// standard stack trace with a more user friendly error message.
+// The calling stack trace is captured before returing the augmented panicHandler
+// The returned panicHandler must be called as a defered function, and must be the
+// first defer called at the start of a new goroutine.
+//
+// This only is a partial solution to the problem of panics within deeply nested
+// go-routines.  It only works between the go-routine being called and the calling
+// go-routine.  If you have multiple nested go-rotuines, it will only preserve the
+// calling stack and the called panic stack.  Idealy we would be able to use context
+// or a similar construct to build a more comprehensive panic handler, but this
+// is a significant step in the right direction that will dramatically improve crash
+// debugging
+func PanicHandlerWithTraceFn() func() {
+	trace := debug.Stack()
```

Comment:
What's the performance impact of this? This would run on every goroutine and collect the stack trace regardless if a panic happened or not.

Reply:
It is called from every "calling" go routine.  The only place that we spawn lots of routines is in the graph walker.  That captures the calling trace before iteration and correctly re-uses the same calling stack.

Reply:
Should we add a comment that callers of this function should be aware that calling this could add a performance impact?

Reply:
Added a comment here and the one place where we are close to a tight loop